lib/inspec/resources/sys_info.rb

```diff
@@ -3,117 +3,182 @@
-                " -f"
-              when "s", "short"
-                " -s"
-              else
-                "ERROR"
-              end
-      end
-      if opt == "ERROR"
-        skip_resource "The `sys_info.hostname` resource is not supported with that option on your OS."
-      else
-        inspec.command("hostname#{opt}").stdout.chomp
-      end
+    def ip_address
+      @cache[:ip]
     end
 
-    # returns the Manufacturer of the local system
     def manufacturer
-      os = inspec.os
-      if os.darwin?
-        "Apple Inc."
-      elsif os.linux?
-        inspec.command("cat /sys/class/dmi/id/sys_vendor").stdout.chomp
-      elsif os.windows?
-        inspec.powershell("Get-CimInstance -ClassName Win32_ComputerSystem | Select Manufacturer -ExpandProperty Manufacturer").stdout.chomp
-      else
-        skip_resource "The `sys_info.manufacturer` resource is not supported on your OS yet."
-      end
+      @cache[:manufacturer]
     end
 
-    # returns the ServerModel of the local system
     def model
-      os = inspec.os
-      if os.darwin?
-        inspec.command("sysctl -n hw.model").stdout.chomp
-      elsif os.linux?
-        inspec.command("cat /sys/class/dmi/id/product_name").stdout.chomp
-      elsif os.windows?
-        inspec.powershell("Get-CimInstance -ClassName Win32_ComputerSystem | Select Model -ExpandProperty Model").stdout.chomp
-      else
-        skip_resource "The `sys_info.model` resource is not supported on your OS yet."
+      @cache[:model]
+    end
+
+    def totalmemory_kb
+      @cache[:totalmemory_kb]
+    end
+
+    def hostname(opt = nil)
+      case opt
+        when nil
+          @cache[:hostname]
+        when "f","long","fqdn","full"
+          @cache[:fqdn]
+        when "d","domain"
+          @cache[:domain]
+        when "i", "ip_address"
+          @cache[:ip_address]
+        when "s","short"
+          @cache[:hostname]
+        else
+          skip_resource "The sys_info.hostname resource is not supported with that option."
       end
     end
 
+    def fqdn
+      @cache[:fqdn]
+    end
+
     def to_s
       "System Information"
     end
   end
+
+  class LinuxInfo < SysInfo
+    attr_reader :inspec
+    def initialize(inspec)
+      @inspec = inspec
+    end
+
+    def info
+      fqdn = inspec.command("hostname -f").stdout.chomp
+      domain = inspec.command("hostname -d").stdout.chomp
+      ip = inspec.command("hostname -I").stdout.chomp
+      short = inspec.command("hostname -s").stdout.chomp
+      manufacturer = inspec.command("cat /sys/class/dmi/id/sys_vendor").stdout.chomp
+      model = inspec.command("cat /sys/class/dmi/id/product_name").stdout.chomp
+      totalmemory_kb = inspec.command("awk '/MemTotal/ {print $2}' /proc/meminfo").stdout.chomp
+
+      {
+        fqdn: fqdn,
+        domain: domain,
+        ip: ip,
+        hostname: short,
+        manufacturer: manufacturer,
+        model: model,
+        totalmemory_kb: totalmemory_kb,
+      }
```

Comment:
Can we construct this structure higher? Doing this in each SysInfo subclass is likely to cause issues when the structure changes in one, but not the other.

Reply:
I think i got your point but moving this into the sys_info class would cause massives changes and will result in different layout as all other ressources: like "services","package". 
But i think i got your point. 
What about updating the methods in root class "sys_info" in this way.

```
    def foobar
      @cache[:foobar] || "The `sys_info.foobar` resource is not supported on your OS yet."
    end
```

In this case the method will return "unsupported" in case the information is not available by the OS subclass.